arkane/encorr/data.py

```diff
@@ -28,10 +28,285 @@
 ###############################################################################
 
 """
-This file provides atomic energy and BAC parameters for several model
+This module provides classes to store data for BAC fitting and
+evaluating.
+
+This file also provides atomic energy and BAC parameters for several model
 chemistries.
 """
 
+import functools
+from collections import Counter
+from dataclasses import dataclass
+from typing import Callable, Iterable, List, Union
+
+import numpy as np
+import pybel
+
+from rmgpy.molecule import Atom, Bond, get_element
+from rmgpy.molecule import Molecule as RMGMolecule
+
+from arkane.encorr.reference import ReferenceSpecies, ReferenceDatabase
+from arkane.exceptions import BondAdditivityCorrectionError
+
+BOND_SYMBOLS = {1: '-', 2: '=', 3: '#'}
+
+
+class Molecule(RMGMolecule):
+    """Wrapper for RMG Molecule to add ID attribute"""
+    def __init__(self, *args, mol_id=None, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.id = mol_id
+
+
+@dataclass
+class Stats:
+    """Small class to store BAC fitting statistics"""
+    rmse: Union[float, np.ndarray]
+    mae: Union[float, np.ndarray]
+
+
+class BACDatapoint:
+    """A BACDatapoint contains a single species"""
+
+    class _Decorators:
+        @staticmethod
+        def assert_model_chemistry(func):
+            @functools.wraps(func)
+            def wrapper(*args, **kwargs):
+                if args[0].model_chemistry is None:  # args[0] is the instance
+                    raise BondAdditivityCorrectionError('Model chemistry is not defined')
+                return func(*args, **kwargs)
+            return wrapper
+
+    def __init__(self, spc: ReferenceSpecies, model_chemistry: str = None):
+        self.spc = spc
+        self.model_chemistry = model_chemistry
+
+        self._mol = None
+        self._mol_type = None
+        self._bonds = None
+        self._ref_data = None
+        self._calc_data = None
+        self._bac_data = None
+        self.weight = 1
```

Comment:
I don't remember exactly why I coded it in this way, but it could be because I was trying to avoid calling `to_mol` unless necessary to avoid the cost associated with it.

Reply:
Oh and also because I don't know if I will need `from_geo` or not when instantiating the class.

Reply:
Ahh okay, I forgot that `from_geo` was an option as well.